README
I wrote this program to avoid the extreemly tedious process of hand creating
stripboards for use with fritzing.  Hopefully other people will find it 
useful as well.
 
It should be straight forward to build on most systems.  You need some version 
of Tcl installed and you need tcllib installed.  The program uses TclKits so 
it is not really dependent on the version of Tcl you have installed, but it 
does use tcllib and generaly tcllib won't install if Tcl itself is not 
installed.  It also uses Doxygen and LaTeX to generate the documentation.  Oh, 
it uses autotools and make (GnuMake) as part of the build process.

It is a purely command line program.  There is no GUI.  To generate a 
stripboard, you will need to write some (simple) Tcl.  On the one hand you can 
use loop structures to generate the bazillian holes.  On the other hand it 
probably means some actually programming and some knowledge of Tcl.  There is 
some "sample" code and I tried to document the API reasonably well, but please 
let me know if there anything that is not clear.

Robert Heller
heller@deepsoft.com


ChangeLog
Wed Sep 25 15:43:19 2019
  1.0.0 Initial Release -- Robert Heller <heller@deepsoft.com>


Scripts/ZipArchive.tcl
#* 
#* ------------------------------------------------------------------
#* Role PlayingDB V3.0 by Deepwoods Software
#* ------------------------------------------------------------------
#* ZipArchive.tcl - Create a Zip Archive using zlib
#* Created by Robert Heller on Sun Aug 30 20:45:47 2009
#* ------------------------------------------------------------------
#* $Id: ZipArchive.tcl 2526 2016-10-17 15:26:20Z heller $
#* ------------------------------------------------------------------
#* Contents:
#* ------------------------------------------------------------------
#*  
#*     Role Playing DB -- A database package that creates and maintains
#* 		       a database of RPG characters, monsters, treasures,
#* 		       spells, and playing environments.
#* 
#*     Copyright (C) 2009  Robert Heller D/B/A Deepwoods Software
#* 			51 Locke Hill Road
#* 			Wendell, MA 01379-9728
#* 
#*     This program is free software; you can redistribute it and/or modify
#*     it under the terms of the GNU General Public License as published by
#*     the Free Software Foundation; either version 2 of the License, or
#*     (at your option) any later version.
#* 
#*     This program is distributed in the hope that it will be useful,
#*     but WITHOUT ANY WARRANTY; without even the implied warranty of
#*     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#*     GNU General Public License for more details.
#* 
#*     You should have received a copy of the GNU General Public License
#*     along with this program; if not, write to the Free Software
#*     Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
#* 
#*  
#* 
#* Lifted from http://wiki.tcl.tk/15158
#* Converted from Artur Trzewik XOTcl code to a snit::type
#*

#@Chapter:ZipArchive.tcl -- Code to create a Zip archive
#$Id: ZipArchive.tcl 2526 2016-10-17 15:26:20Z heller $
# This is a port of Artur Trzewik's code that creates a Zip archive
# written in XOTcl to a snit::type.  It has been modified to handle
# directories and includes code to recursively traverse a directory.
#

package require vfs::zip
package require vfslib
package require snit

snit::type ZipArchive {
  variable files {}
  variable cdOffset 0 
  variable cdLength 0
  variable written 0
  method addFile {inputFile fileName} {
    lappend files $inputFile $fileName
  }

  method addToStream {stream fin fout} {
    set offset $written
    if {[file isdirectory $fin]} {
      set data ""
      append fout "/"
      set datacompresed [string range [::vfs::zip -mode compress $data] 2 end-4]
    } else {
      set fdata [open $fin r]
      fconfigure $fdata -encoding binary -translation binary
      set data [read $fdata]
      set datacompresed [string range [::vfs::zip -mode compress $data] 2 end-4]
      close $fdata
    }

    binary scan \x04\x03\x4B\x50 I LFH_SIG
    $self writeLong $stream $LFH_SIG

    incr written 4

    $self writeShort $stream 20
    # java implementation make 8
    # but tools (WinZip) leave it 0
    $self writeShort $stream 0
    incr written 4

    $self writeShort $stream 8
    incr written 2

    # last mod. time and date
    set dosTime [$self toDosTime $fin]

    $self writeLong $stream $dosTime
    incr written 4

    set crc [::vfs::crc $data]
    set csize [string length $datacompresed]
    set size [string length $data]

    $self writeLong $stream $crc
    $self writeLong $stream $csize
    $self writeLong $stream $size
    incr written 12

    # file name length
    $self writeShort $stream [string length $fout]
    incr written 2

    # extra field length
    set extra ""
    $self writeShort $stream [string length $extra]
    incr written 2

    # file name
    puts -nonewline $stream $fout
    incr written [string length $fout]

    puts -nonewline $stream $extra
    incr written [string length $extra]

    set dataStart $written
    puts -nonewline $stream $datacompresed
    incr written $csize

    return [list $offset $dosTime $crc $csize $size]
  }

  method createFile {file args} {
    set fout [open $file w]
    fconfigure $fout -encoding binary -translation binary
    eval [list $self createToStream $fout] $args
    close $fout
  }

  method createToStream {ostream args} {
    set descriptionList [list]
    foreach {fin fout} $files {
      lappend descriptionList [$self addToStream $ostream $fin $fout]
    }
    set cdOffset $written

    foreach {fin fout} $files desc $descriptionList {
      foreach {offset dosTime crc csize size} $desc {}
      $self writeCentralFileHeader $ostream $fin $fout $offset $dosTime $size $csize $crc
    }

    set cdLength [expr {$written - $cdOffset}]
    # wirte header

    # EOCD 0X06054B50L scan 0X06054B50L %x s set s
    binary scan \x06\x05\x4B\x50 I EOCD
    $self writeLong $ostream $EOCD

    # disk numbers
    $self writeShort $ostream 0
    $self writeShort $ostream 0

    # number of entries
    set filenum [expr {[llength $files]>>1}]
    $self writeShort $ostream $filenum
    $self writeShort $ostream $filenum

    # length and location of CD
    $self writeLong $ostream $cdLength
    $self writeLong $ostream $cdOffset

    # zip file comment
    set comment [from args -comment ""]
    # comment lenght
    $self writeShort $ostream [string bytelength $comment]

    puts -nonewline $ostream $comment
  }
  constructor {} {
    set files [list]
    set cdLength 0
    set cdOffset 0
    set written 0
  }
  method toDosTime {file} {
    set sec [file mtime $file]
    if {$sec == 0} {;#		broken metakit VFS directory implementation???
      set sec [clock scan now];# temp hack to keep the vfs::zip implementation happy.
    }

    foreach {year month day hour minute secound} \
	[clock format $sec -format "%Y %m %e %k %M %S"] {}

    set month [string trimleft $month 0]
    set year [string trimleft $year 0]
    set minute [string trimleft $minute 0]
    if {$minute eq ""} {
      set minute 0
    }
    set secound [string trimleft $secound 0]
    if {$secound eq ""} {
      set secound 0
    }

    set hour  [string trimleft $hour]
    if {$hour eq ""} {set hour 0}
    set value [expr {(($year - 1980) << 25) | \
		     ($month << 21) | ($day << 16) | ($hour << 11) | \
		     ($minute << 5) | ($secound >> 1)}]
    return $value
  }
  method writeCentralFileHeader {ostream fin fout offset dosTime size csize crc} {
    # CFH 0X02014B50L
    binary scan \x02\x01\x4B\x50 I CFH_SIG
    $self writeLong $ostream $CFH_SIG
    incr written 4

    if {$::tcl_platform(platform) eq "windows"} {
        # unix
        set pid 5
    } else {
        # windows
        set pid 11
    }
    $self writeShort $ostream [expr { (($pid << 8) | 20)}]
    incr written 2

    # version needed to extract
    # general purpose bit flag

    $self writeShort $ostream 20
    $self writeShort $ostream 0
    incr written 4

    # compression method
    $self writeShort $ostream 8
    incr written 2

    # last mod. time and date
    $self writeLong $ostream $dosTime
    incr written 4

    # CRC
    # compressed length
    # uncompressed length
    $self writeLong $ostream $crc
    $self writeLong $ostream $csize
    $self writeLong $ostream $size
    incr written 12;

    set comment ""
    set extra ""

    # file name length

    $self writeShort $ostream [string bytelength $fout]
    incr written 2;

    # extra field length
    $self writeShort $ostream [string bytelength $extra]
    incr written 2;

    # file comment length
    $self writeShort $ostream [string bytelength $comment]
    incr written 2;

    # disk number start
    $self writeShort $ostream 0
    incr written 2

    # internal file attributes
    $self writeShort $ostream 0
    incr written 2

    # external file attributes
    $self writeLong $ostream 0
    incr written 4

    # relative offset of LFH
    $self writeLong $ostream $offset
    incr written 4

    # file name
    puts -nonewline $ostream $fout
    incr written [string bytelength $fout]

    # extra field
    puts -nonewline $ostream $extra
    incr written [string bytelength $extra]

    # file comment
    puts -nonewline $ostream $comment
    incr written [string bytelength $comment]
  }
  method writeLong {stream short} {
    puts -nonewline $stream [binary format i $short]
  }
  method writeShort {stream short} {
    puts -nonewline $stream [binary format s $short]
  }
  proc find {directory} {
    set result [list]
    foreach file [glob -nocomplain [file join $directory *]] {
      if {[file isdirectory $file]} {
	eval [list lappend result $file/] [find $file]
      } else {
	lappend result $file
      }
    }
    return $result
  }
  proc stripdir {file directory} {
    set dlen [string length $directory]; incr dlen -1
    if {"[string range $file 0 $dlen]" eq "$directory"} {
      return [string range "$file" [expr {$dlen + 2}] end]
    } else {
      return "$file"
    }
  }
  typemethod createZipFromDirtree {zipfile directory args} {
  # This type method creates a Zip archive from a directory tree.
  #
  # Args:
  # <in> zipfile The name of the zip file to create.
  # <in> directory The name of the directory to recursively traverse.
  # <in> args Any remaining args to pass to the createFile method.

    set ziparchive [$type create %AUTO%]
    foreach f [find $directory] {
      $ziparchive addFile $f [stripdir $f $directory]
    }
    eval [list $ziparchive createFile $zipfile] $args
    $ziparchive destroy
    return $zipfile
  }
}

package provide ZipArchive 1.0







    


BoardScripts/BPS_POW3U.tcl
#*****************************************************************************
#
#  System        : 
#  Module        : 
#  Object Name   : $RCSfile$
#  Revision      : $Revision$
#  Date          : $Date$
#  Author        : $Author$
#  Created By    : Robert Heller
#  Created       : Wed Sep 25 18:34:39 2019
#  Last Modified : <190925.2001>
#
#  Description	
#
#  Notes
#
#  History
#	
#*****************************************************************************
#
#    Copyright (C) 2019  Robert Heller D/B/A Deepwoods Software
#			51 Locke Hill Road
#			Wendell, MA 01379-9728
#
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
#
# 
#
#*****************************************************************************


snit::type BPS_POW3U {
    # Use the macro to include the component
    Board "BusBoard Prototype Systems BPS POW3U Stripboard"
    proc _letterColumn {column} {
        if {$column >= 1 && $column <= 26} {
            #puts stderr "*** _letterColumn: $column is between 1 and 26"
            return [format {%c} [expr {64 + $column}]]
        } elseif {$column >= 27 && $column <= 52} {
            #puts stderr "*** _letterColumn: $column is between 27 and 52"
            return [format {%c} [expr {96 + ($column - 26)}]]
        } else {
            #puts stderr "*** _letterColumn: $column is greater than 52"
            set ch [expr {64 + ($column - 52)}]
            return [format {%c%c} $ch $ch]
        }
    }
    constructor {args} {
        # Install the component, customizing the metadata.
        install board using GenericStripboard ${selfns}%AUTO% \
              -version 1.0 \
              -author "Robert Heller" \
              -title "BPS POW3U" \
              -moduleid "BPS_POW3U" \
              -label POW3U \
              -description {* PowerBoard has interleaved power and
            ground rails to easily distribute power to
            your circuits.
        
            * The general purpose strip board circuit
            pattern with 6 holes per strip is good for
            general purpose analog and digital use.
        
            * 0.1" hole spacing for DIP integrated circuits
            and headers.
        
            * Standard single height (3U) Eurocard/VME
            size. Many off the shelf enclosures and card
            cages are available
        
            * Accepts a 96 pin DIN-41612 VME
            connector for backplane or board-to-board
            connections. Holes for ejector latches.
            Rows 1 and 3 are routed to separate pads.
            Row 2 is unconnected.
        
            * Drilled holes for ejector latches.} \
            -units mm -width 160 -height 100 \
            -viewport {0 0 16000 10000}
        # Mounting and board ejector holes (non-connector holes).
        $board AddHole 350 508 76.2
        $board AddHole 350 9492 76.2
        $board AddHole 15750 508 200
        $board AddHole 15750 9492 200
        # Vertical strip at the back of the board
        for {set y 762;set r 3} {$r <= 36} {incr y 254;incr r} {
            $board AddHole 175 $y 47 ${r}A A
        }
        $board AddLine 175 762 175 [expr {$y - 254}] brown 150 "stroke-linecap:round;stroke-opacity:.5;"
                # Now do the batches of horizontal strips
        set connCol 1
        set y 0
        for {set row 1} {$row <= 38} {incr row} {
            set x [expr {175 + 254}]
            incr y 254
            # The first and last row are 60 hole strips
            if {$row == 1 || $row == 38} {
                incr x 254
                set x1 $x
                for {set b 3} {$b < 62} {incr b} {
                    set col [_letterColumn $b]
                    $board AddHole $x $y 47 ${row}${col} $row
                    set x2 $x
                    incr x 254
                }
                $board AddLine $x1 $y $x2 $y brown 150 "stroke-linecap:round;stroke-opacity:.5;"
            } elseif {$row == 2 || $row == 37} {
                # The second and next to last are 59 hole strips and part of 
                # the power / ground buses
                incr x 254
                set x1 $x
                if {$row == 2} {
                    set ypA1 $y
                    #set ypB1 [expr {$y + 254}]
                    set bus P1
                } else {
                    #set ypA2 [expr {$y - 254}]
                    set ypB2 $y
                    set bus P2
                }
                set x1 $x
                for {set b 3} {$b < 61} {incr b} {
                    set col [_letterColumn $b]
                    $board AddHole $x $y 47 ${row}${col} $bus
                    set x2 $x
                    incr x 254
                }
                $board AddLine $x1 $y $x2 $y brown 150 "stroke-linecap:round;stroke-opacity:.5;"
            } elseif {$row == 3 || $row == 36} {
                # The third and third from the last have eight six hole buses, 
                # four pairs of vertical power buses and four hole bus
                if {$row == 3} {
                    set ypB1 $y
                } else {
                    set ypA2 $y
                }
                set pIndex 0
                for {set b 1} {$b <= 8} {incr b} {
                    set col [_letterColumn $b]
                    set x1 $x
                    set bus ${row}${col}
                    for {set i 1} {$i <= 6} {incr i} {
                        $board AddHole $x $y 47 ${bus}$i $bus
                        set x2 $x
                        incr x 254
                    }
                    $board AddLine $x1 $y $x2 $y brown 150 "stroke-linecap:round;stroke-opacity:.5;"
                    if {($b&1) == 1} {
                        $board AddHole $x $y 47 ${row}P1_$pIndex P1
                        set xpA($pIndex) $x
                        incr x 254
                        $board AddHole $x $y 47 ${row}P2_$pIndex P2
                        set xpB($pIndex) $x
                        incr x 254
                        incr pIndex
                    }
                }
                set col [_letterColumn $b]
                set x1 $x
                set bus ${row}${col}
                for {set i 1} {$i <= 4} {incr i} {
                    $board AddHole $x $y 47 ${bus}$i $bus
                    set x2 $x
                    incr x 254
                }
                $board AddLine $x1 $y $x2 $y brown 150 "stroke-linecap:round;stroke-opacity:.5;"
            } else {
                # And the rest have the DIN-41612 connector...
                set pIndex 0
                for {set b 1} {$b <= 7} {incr b} {
                    set col [_letterColumn $b]
                    set x1 $x
                    set bus ${row}${col}
                    for {set i 1} {$i <= 6} {incr i} {
                        $board AddHole $x $y 47 ${bus}$i $bus
                        set x2 $x
                        incr x 254
                    }
                    $board AddLine $x1 $y $x2 $y brown 150 "stroke-linecap:round;stroke-opacity:.5;"
                    if {($b&1) == 1} {
                        $board AddHole $x $y 47 ${row}P1_$pIndex P1
                        incr x 254
                        $board AddHole $x $y 47 ${row}P2_$pIndex P2
                        incr x 254
                        incr pIndex
                    }
                }
                set bus 3C$connCol
                set x1 $x
                for {set i 1} {$i <= 6} {incr i} {
                    $board AddHole $x $y 47 ${bus}[_letterColumn $i] $bus
                    set x2 $x
                    set C3_x1 $x
                    incr x 254
                }
                $board AddLine $x1 $y $x2 $y brown 150 "stroke-linecap:round;stroke-opacity:.5;"
                $board AddHole $x $y 44.45 1C${connCol}A 1C$connCol
                set x1 $x
                incr x 254 
                $board AddHole $x $y 44.45 1C${connCol}B 1C$connCol
                set x2 $x
                $board AddLine $x1 $y $x2 $y brown 150 "stroke-linecap:round;stroke-opacity:.5;"
                incr x 254
                $board AddHole $x $y 44.45 2C${connCol} 2C$connCol
                incr x 254
                $board AddHole $x $y 44.45 3C${connCol}G 3C$connCol
                $board AddLine $x $y $x $y brown 150 "stroke-linecap:round;stroke-opacity:.5;"
                set C3_x2 $x
                set C3_y [expr {$y - 127}]
                #puts stderr "*** $type create $self: C3_x1 = $C3_x1, C3_y = $C3_y, C3_x2 = $C3_x2"
                $board AddLine $C3_x1 $C3_y $C3_x2 $C3_y brown 30 "stroke-opacity:.5;"
                $board AddLine $C3_x1 $C3_y $C3_x1 $y brown 30 "stroke-linecap:butt;stroke-opacity:.5;"
                $board AddLine $C3_x2 $C3_y $C3_x2 $y brown 30 "stroke-linecap:butt;stroke-opacity:.5;"
                incr x 254
                incr connCol
            }
        }
        foreach pIndex [array names xpA] {
            $board AddLine $xpA($pIndex) $ypA1 $xpA($pIndex) $ypA2 brown 150 "stroke-linecap:round;stroke-opacity:.5;"
        }
        foreach pIndex [array names xpB] {
            $board AddLine $xpB($pIndex) $ypB1 $xpB($pIndex) $ypB2 brown 150 "stroke-linecap:round;stroke-opacity:.5;"
        }
    }
}

set b [BPS_POW3U create %AUTO%]
$b WriteFZPZ [$b cget -moduleid].fzpz

        


BoardScripts/BPS_BR1.tcl
#*****************************************************************************
#
#  System        : 
#  Module        : 
#  Object Name   : $RCSfile$
#  Revision      : $Revision$
#  Date          : $Date$
#  Author        : $Author$
#  Created By    : Robert Heller
#  Created       : Fri Sep 27 18:48:13 2019
#  Last Modified : <190927.2114>
#
#  Description	
#
#  Notes
#
#  History
#	
#*****************************************************************************
#
#    Copyright (C) 2019  Robert Heller D/B/A Deepwoods Software
#			51 Locke Hill Road
#			Wendell, MA 01379-9728
#
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
#
# 
#
#*****************************************************************************


snit::type BPS_BR1 {
    # Use the macro to include the component
    Board "BusBoard Prototype Systems BPS BR1 Solderable PC BreadBoard"
    constructor {args} {
        # Install the component, customizing the metadata.
        install board using GenericStripboard ${selfns}%AUTO% \
              -version 1.0 \
              -author "Robert Heller" \
              -title "BPS BR1" \
              -moduleid "BPS_BR1" \
              -label BR1 \
              -description {* Transfer your circuit and wires without
            recutting wires or changing your layout.
            
            The BR1 Solderable BreadBoard has the
            same pattern and spacing as a standard 830
            connection point solderless breadboard.
            
            * Soldered joints provide a more reliable
            longterm connection than solderless
            contacts.
            
            * Standard 5-hole strip for each IC pin, just
            like solderless breadboards.
            
            * Four power rails like solderless breadboards
            plus two bonus rails for power or signals.
            
            * Bonus two-hole pads along the centerline
            allow DIL (dual in-line) headers to be used.
            
            * Lead free etched copper PCB with an
            anti-tarnish coating for easy soldering.
            
            * 0.1" hole spacing for DIP integrated circuits
            and headers.
            
            * Four PCB mounting holes provided.} \
            -units mm -width 179 -height 47 \
            -viewport {0 0 17872 4700}
        # Mounting holes
        $board AddHole 300 445 175
        $board AddHole 17572 445 175
        $board AddHole 17572 4255 175
        $board AddHole 300 4255 175
        for {set col 1;set x 300} {$col < 70} {incr col;incr x 254} {
            # Pad-per-hole holes at the ends
            if {$col == 1 || $col == 69} {
                # Column 1/69
                for {set r 4;set y [expr {445 + 508}]} {$r < 16} {incr r;incr y 254} {
                    $board AddHole $x $y 53.5 [row2letter $r]$col
                }
            } elseif {$col == 2 || $col == 68} {
                # Column 2/68
                    for {set r 3; set y [expr {445 + 254}]} {$r < 17} {incr r;incr y 254} {
                    $board AddHole $x $y 53.5 [row2letter $r]$col
                }
            } elseif {$col == 3  || $col == 4 ||
                      $col == 66 ||$col == 67} {
                # Column 3/66 -- start/end of strips (with longest power strips)
                if {$col == 3} {
                    set p1_x1 $x
                    set p6_x1 $x
                } elseif {$col == 67} {
                    set p1_x2 $x
                    set p6_x2 $x
                }
                for {set r 1;set y [expr {445 - 254}]} {$r < 19} {incr r;incr y 254} {
                    if {$r == 1} {
                        set bus P1
                        set p1_y $y
                    } elseif {$r == 2} {
                        set bus A${col}A
                        set y1 $y
                    } elseif {$r == 8} {
                        set y2 $y
                    } elseif {$r == 9} {
                        set y1 $y
                        set bus A${col}B
                    } elseif {$r == 10} {
                        set y2 $y
                    } elseif {$r == 11} {
                        set y1 $y
                        set bus A${col}C
                    } elseif {$r == 17} {
                        set y2 $y
                    } elseif {$r == 18} {
                        set bus P6
                        set p6_y $y
                    }
                    $board AddHole $x $y 53.5 [row2letter $r]${col} $bus
                    if {$r == 8} {
                        $board AddLine $x $y1 $x $y2 brown 150 "stroke-linecap:round;stroke-opacity:.5;"
                    } elseif {$r == 10} {
                        $board AddLine $x $y1 $x $y2 brown 150 "stroke-linecap:round;stroke-opacity:.5;"
                    } elseif {$r == 17} {
                        $board AddLine $x $y1 $x $y2 brown 150 "stroke-linecap:round;stroke-opacity:.5;"
                    }
                }
            } else {
                # Columns 5 though 65
                if {$col == 5} {
                    set p2_x1 $x
                    set p3_x1 $x
                    set p4_x1 $x
                    set p5_x1 $x
                } elseif {$col == 65} {
                    set p2_x2 $x
                    set p3_x2 $x
                    set p4_x2 $x
                    set p5_x2 $x
                }
                for {set r 1;set y [expr {445 - 254}]} {$r < 19} {incr r;incr y 254} {
                    if {$r == 1} {
                        set bus P1
                        set p1_y $y
                    } elseif {$r == 2} {
                        set bus P2
                        set p2_y $y
                    } elseif {$r == 3} {
                        set bus P3
                        set p3_y $y
                    } elseif {$r == 4} {
                        set bus A${col}A
                        set y1 $y
                    } elseif {$r == 8} {
                        set y2 $y
                    } elseif {$r == 9} {
                        set y1 $y
                        set bus A${col}B
                    } elseif {$r == 10} {
                        set y2 $y
                    } elseif {$r == 11} {
                        set y1 $y
                        set bus A${col}C
                    } elseif {$r == 15} {
                        set y2 $y
                    } elseif {$r == 16} {
                        set bus P4
                        set p4_y $y
                    } elseif {$r == 17} {
                        set bus P5
                        set p5_y $y
                    } elseif {$r == 18} {
                        set bus P6
                        set p6_y $y
                    }
                    $board AddHole $x $y 53.5 [row2letter $r]${col} $bus
                    if {$r == 8} {
                        $board AddLine $x $y1 $x $y2 brown 150 "stroke-linecap:round;stroke-opacity:.5;"
                    } elseif {$r == 10} {
                        $board AddLine $x $y1 $x $y2 brown 150 "stroke-linecap:round;stroke-opacity:.5;"
                    } elseif {$r == 15} {
                        $board AddLine $x $y1 $x $y2 brown 150 "stroke-linecap:round;stroke-opacity:.5;"
                    }
                }
            }
        }
        #puts stderr "*** $type create $self: x = $x"
        $board AddLine $p1_x1 $p1_y $p1_x2 $p1_y brown 150 "stroke-linecap:round;stroke-opacity:.5;"
        $board AddLine $p2_x1 $p2_y $p2_x2 $p2_y brown 150 "stroke-linecap:round;stroke-opacity:.5;"
        $board AddLine $p3_x1 $p3_y $p3_x2 $p3_y brown 150 "stroke-linecap:round;stroke-opacity:.5;"
        $board AddLine $p4_x1 $p4_y $p4_x2 $p4_y brown 150 "stroke-linecap:round;stroke-opacity:.5;"
        $board AddLine $p5_x1 $p5_y $p5_x2 $p5_y brown 150 "stroke-linecap:round;stroke-opacity:.5;"
        $board AddLine $p6_x1 $p6_y $p6_x2 $p6_y brown 150 "stroke-linecap:round;stroke-opacity:.5;"
    }
    proc row2letter {row} {
        return [format {%c} [expr {$row + 64}]]
    }
}

set b [BPS_BR1 create %AUTO%]
$b WriteFZPZ [$b cget -moduleid].fzpz
#$b WriteBBSVG [$b cget -label].svg

            
